Issue/8690 site creation titles

malinajirka opened the pull request:
Fixes #8690 

Adds correct screen titles to the steps in NewSiteCreation flow
- "Create Site" on the first step
- "Empty title" on the last step
- "n of x" on the remaining steps

To test:
- update`NEW_SITE_CREATION_ENABLED` in build.gradle
- Go to `My Site` -> `Switch Site` -> plus sign icon in the top right corner
- check the title of the Segments screen is "Create Site"
- select a segment
- check the title of the Verticals screen is "1 of 3"
- type something and select a vertical
- check the title of the Site Info screen is "2 of 3"
- click on the skip button
- check the title of the Domains screen is "3 of 3"
- press back button and check the title of the Site Info screen is "2 of 3"
- enter a SiteTitle and click on the next button
- check the title of the Domains screen is "3 of 3"

Bonus: you can test the title is being read by TalkBack when you enter or re-enter each step screen or we can leave it for the #8510

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSiteCreationActivity.kt:
```diff
@@ -60,14 +62,29 @@ class NewSiteCreationActivity : AppCompatActivity(),
     private fun showStep(target: WizardNavigationTarget<SiteCreationStep, SiteCreationState>) {
         val fragment = when (target.wizardStep) {
-            SEGMENTS -> NewSiteCreationSegmentsFragment.newInstance()
+            SEGMENTS -> NewSiteCreationSegmentsFragment.newInstance(getScreenTitle(target.wizardStep))
```
Can we introduce a local `val` for `getScreenTitle(target.wizardStep)` instead of repeating in each case?

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSiteCreationBaseFormFragment.java:
```diff
@@ -48,39 +50,49 @@ public void onCreate(Bundle savedInstanceState) {
         if (actionBar != null) {
             actionBar.setDisplayHomeAsUpEnabled(true);
-            actionBar.setTitle(R.string.site_creation_title);
+            actionBar.setTitle(getScreenTitle());
```
I am really not a big fan of doing this in the base fragment. If the instance creation went through the base fragment, I'd be fine with it, but since we have the `newInstance` in each fragment and the titles are set there, I think I'd have preferred to have it in each fragment even though it's a bit of duplication.

If you feel strongly about keeping it as is, I don't mind it **that much**, so feel free to ignore this comment.

malinajirka replied:
Since I want to force the children to set both actionbar title and activity title I made the getScreenTitle method abstract. I hope it makes it cleaner. Let me know if you want me to modify something else.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/verticals/NewSiteCreationVerticalsFragment.kt:
```diff
@@ -81,8 +81,6 @@ class NewSiteCreationVerticalsFragment : NewSiteCreationBaseFormFragment<NewSite
 
     override fun setupContent(rootView: ViewGroup) {
         // TODO receive title from the MainVM
```
Can we remove this comment now?

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/util/wizard/WizardManager.kt:
```diff
@@ -25,6 +25,19 @@ class WizardManager<T : WizardStep>(
         return currentStepIndex >= 0 && currentStepIndex < steps.size
     }
+
+    fun stepsCount(): Int = steps.size
```
We can make this a `val` instead.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/util/wizard/WizardManager.kt:
```diff
@@ -25,6 +25,19 @@ class WizardManager<T : WizardStep>(
+        return if (steps.contains(T)) {
+            steps.indexOf(T) + 1
+        } else {
+            -1
```
Can we throw an exception when this happens or at least return `null` if it's a valid state we need to handle in runtime? If the concern here is that Java won't warn us for the `null` value, it's the case for `-1` as well. Could you elaborate on this decision?

malinajirka replied:
I'm not sure what I was thinking about. I guess I'm just used to return -1 when working with indices. Fixed in [270e831](https://github.com/wordpress-mobile/WordPress-Android/pull/8756/commits/270e831bf195de50a3f826ac19b6524fe044b85e)

malinajirka commented:
Thanks @oguzkocer. I've fixed all the issues.

I've also merged the master-site-creation branch and I fixed the logic of titles when the last step of the wizard (Site Preview) is present. ( [8e5f94c](https://github.com/wordpress-mobile/WordPress-Android/pull/8756/commits/8e5f94c1b951189d5c30eccd97a1cd940f31be08)) I updated the Test steps in the description of this PR.
